sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,769 @@
+
+    maxima_ratio : float, optional
+        The maximum ratio we allow of average height of clusters on the
+        right and left to the local maxima in question. The higher the
+        ratio, the more generous the algorithm is to preserving local
+        minima, and the more cuts the resulting tree will have.
+
+    rejection_ratio : float, optional
+        Adjusts the fitness of the clustering. When the maxima_ratio is
+        exceeded, determine which of the clusters to the left and right to
+        reject based on rejection_ratio. Higher values will result in points
+        being more readily classified as noise; conversely, lower values will
+        result in more points being classified.
+
+    similarity_threshold : float, optional
+        Used to check if nodes can be moved up one level, that is, if the
+        new cluster created is too "similar" to its parent, given the
+        similarity threshold. Similarity can be determined by 1) the size
+        of the new cluster relative to the size of the parent node or
+        2) the average of the reachability values of the new cluster
+        relative to the average of the reachability values of the parent
+        node. A lower value for the similarity threshold means less levels
+        in the tree.
+
+    significant_min : float, optional
+        Sets a lower threshold on how small a significant maxima can be
+
+    min_cluster_size_ratio : float, optional
+        Minimum percentage of dataset expected for cluster membership.
+
+    min_maxima_ratio : float, optional
+        Used to determine neighborhood size for minimum cluster membership.
+
+    Returns
+    -------
+    core_sample_indices_ : array, shape (n_core_samples,)
+        The indices of the core samples.
+
+    labels_ : array, shape (n_samples,)
+        The estimated labels.
+    """
+
+    # Extraction wrapper
+    reachability_plot = reachability[ordering].tolist()
+    root_node = _automatic_cluster(reachability_plot, ordering,
+                                   maxima_ratio, rejection_ratio,
+                                   similarity_threshold, significant_min,
+                                   min_cluster_size_ratio, min_maxima_ratio)
+    leaves = _get_leaves(root_node, [])
+    # Start cluster id's at 1
+    clustid = 0
+    n_samples = len(reachability)
+    is_core = np.zeros(n_samples, dtype=bool)
+    labels = np.ones(n_samples, dtype=int) * -1
+    # Start all points as non-core noise
+    for leaf in leaves:
+        index = ordering[leaf.start:leaf.end]
+        labels[index] = clustid
+        is_core[index] = 1
+        clustid += 1
+    return(np.arange(n_samples)[is_core], labels)
+
+
+def _automatic_cluster(reachability_plot, reachability_ordering,
+                       maxima_ratio, rejection_ratio,
+                       similarity_threshold, significant_min,
+                       min_cluster_size_ratio, min_maxima_ratio):
```

Comment:
can you add a simple docstring?

Reply:
added

Reply:
I think the request here was for a description of parameters and types that might not be patently obvious

Reply:
At least, you should describe the parameters that differ from `_extract_optics`, being `reachability_plot` and `reachability_ordering`